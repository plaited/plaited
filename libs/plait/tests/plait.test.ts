import {
  assertEquals,
  assertSnapshot,
  beforeEach,
  describe,
  it,
} from '../../test-deps.ts'
import {
  block,
  chaosStrategy,
  loop,
  plait,
  randomizedStrategy,
  request,
  strand,
  waitFor,
} from '../mod.ts'

describe('Plait', () => {
  const actualFeedback: string[] = []
  const expectedFeedback = [
    'Add hot',
    'Add cold',
    'Add hot',
    'Add cold',
    'Add hot',
    'Add cold',
  ]
  beforeEach(() => {
    actualFeedback.length = 0
  })

  const addHot = () => actualFeedback.push('Add hot')
  const addCold = () => actualFeedback.push('Add cold')
  const strands = {
    addHot: strand(
      waitFor<string>({
        type: 'start',
        assert({ data }) {
          return data === 'start'
        },
      }),
      request({ type: 'hot' }),
      request({ type: 'hot' }),
      request({ type: 'hot' }),
    ),
    addCold: strand(
      waitFor({ type: 'start' }),
      request({ type: 'cold' }),
      request({ type: 'cold' }),
      request({ type: 'cold' }),
    ),
    mixHotCold: loop(
      strand(
        Object.assign(
          waitFor({ type: 'hot' }),
          block({ type: 'cold' }),
        ),
        Object.assign(
          waitFor({ type: 'cold' }),
          block({ type: 'hot' }),
        ),
      ),
    ),
  }
  const actions = {
    cold() {
      addCold()
    },
    hot() {
      addHot()
    },
  }
  it('priority queue', (t) => {
    const streamLog: unknown[] = []
    const { trigger, feedback, stream, add } = plait({ dev: true })
    add(strands)
    feedback(actions)
    stream.subscribe((msg) => {
      streamLog.push(msg)
    })
    trigger({
      type: 'start',
      data: 'start',
    })
    assertEquals(
      actualFeedback,
      expectedFeedback,
      `priority selection feedback`,
    )
    assertSnapshot(t, streamLog, `priority selection feedback`)
  })
  it('randomized priority queue', (t) => {
    const streamLog: unknown[] = []
    const { trigger, feedback, stream, add } = plait({
      strategy: randomizedStrategy,
      dev: true,
    })
    add(strands)
    feedback(actions)
    stream.subscribe((msg) => {
      streamLog.push(msg)
    })
    trigger({
      type: 'start',
      data: 'start',
    })
    assertEquals(
      actualFeedback,
      expectedFeedback,
      `randomized priority selection feedback`,
    )
    assertSnapshot(t, streamLog, `randomized priority selection log`)
  })
  it('chaos selection', (t) => {
    const streamLog: unknown[] = []
    const { trigger, feedback, stream, add } = plait({
      strategy: chaosStrategy,
      dev: true,
    })
    add(strands)
    feedback(actions)
    stream.subscribe((msg) => {
      streamLog.push(msg)
    })
    trigger({
      type: 'start',
      data: 'start',
    })
    assertEquals(actualFeedback, expectedFeedback, `chaos selection feedback`)
    assertSnapshot(t, streamLog, `chaos selection log`)
  })
})
