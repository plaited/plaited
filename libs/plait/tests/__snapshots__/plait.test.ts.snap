export const snapshot = {};

snapshot[`Plait > priority queue 1`] = `
[
  {
    detail: {
      data: "start",
      type: "start",
    },
    type: "trigger-event",
  },
  {
    detail: {
      blockedEvents: [
        "cold",
      ],
      logicStrands: [
        {
          priority: 1,
          strandName: "addHot",
        },
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
        {
          priority: 0,
          strandName: "start",
        },
      ],
      requestedEvents: [
        {
          data: "start",
          priority: 0,
          type: "start",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      data: "start",
      type: "start",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "cold",
      ],
      logicStrands: [
        {
          priority: 3,
          strandName: "mixHotCold",
        },
        {
          priority: 1,
          strandName: "addHot",
        },
        {
          priority: 2,
          strandName: "addCold",
        },
      ],
      requestedEvents: [
        {
          priority: 1,
          type: "hot",
        },
        {
          priority: 2,
          type: "cold",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "hot",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "hot",
      ],
      logicStrands: [
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
        {
          priority: 1,
          strandName: "addHot",
        },
      ],
      requestedEvents: [
        {
          priority: 2,
          type: "cold",
        },
        {
          priority: 1,
          type: "hot",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "cold",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "cold",
      ],
      logicStrands: [
        {
          priority: 1,
          strandName: "addHot",
        },
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
      ],
      requestedEvents: [
        {
          priority: 1,
          type: "hot",
        },
        {
          priority: 2,
          type: "cold",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "hot",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "hot",
      ],
      logicStrands: [
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 1,
          strandName: "addHot",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
      ],
      requestedEvents: [
        {
          priority: 2,
          type: "cold",
        },
        {
          priority: 1,
          type: "hot",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "cold",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "cold",
      ],
      logicStrands: [
        {
          priority: 1,
          strandName: "addHot",
        },
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
      ],
      requestedEvents: [
        {
          priority: 1,
          type: "hot",
        },
        {
          priority: 2,
          type: "cold",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "hot",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "hot",
      ],
      logicStrands: [
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
      ],
      requestedEvents: [
        {
          priority: 2,
          type: "cold",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "cold",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "cold",
      ],
      logicStrands: [
        {
          priority: 3,
          strandName: "mixHotCold",
        },
      ],
      requestedEvents: [
      ],
    },
    type: "state-snapshot",
  },
]
`;

snapshot[`Plait > randomized priority queue 1`] = `
[
  {
    detail: {
      data: "start",
      type: "start",
    },
    type: "trigger-event",
  },
  {
    detail: {
      blockedEvents: [
        "cold",
      ],
      logicStrands: [
        {
          priority: 1,
          strandName: "addHot",
        },
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
        {
          priority: 0,
          strandName: "start",
        },
      ],
      requestedEvents: [
        {
          data: "start",
          priority: 0,
          type: "start",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      data: "start",
      type: "start",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "cold",
      ],
      logicStrands: [
        {
          priority: 3,
          strandName: "mixHotCold",
        },
        {
          priority: 1,
          strandName: "addHot",
        },
        {
          priority: 2,
          strandName: "addCold",
        },
      ],
      requestedEvents: [
        {
          priority: 1,
          type: "hot",
        },
        {
          priority: 2,
          type: "cold",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "hot",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "hot",
      ],
      logicStrands: [
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
        {
          priority: 1,
          strandName: "addHot",
        },
      ],
      requestedEvents: [
        {
          priority: 2,
          type: "cold",
        },
        {
          priority: 1,
          type: "hot",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "cold",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "cold",
      ],
      logicStrands: [
        {
          priority: 1,
          strandName: "addHot",
        },
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
      ],
      requestedEvents: [
        {
          priority: 1,
          type: "hot",
        },
        {
          priority: 2,
          type: "cold",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "hot",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "hot",
      ],
      logicStrands: [
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 1,
          strandName: "addHot",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
      ],
      requestedEvents: [
        {
          priority: 2,
          type: "cold",
        },
        {
          priority: 1,
          type: "hot",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "cold",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "cold",
      ],
      logicStrands: [
        {
          priority: 1,
          strandName: "addHot",
        },
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
      ],
      requestedEvents: [
        {
          priority: 1,
          type: "hot",
        },
        {
          priority: 2,
          type: "cold",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "hot",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "hot",
      ],
      logicStrands: [
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
      ],
      requestedEvents: [
        {
          priority: 2,
          type: "cold",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "cold",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "cold",
      ],
      logicStrands: [
        {
          priority: 3,
          strandName: "mixHotCold",
        },
      ],
      requestedEvents: [
      ],
    },
    type: "state-snapshot",
  },
]
`;

snapshot[`Plait > chaos selection 1`] = `
[
  {
    detail: {
      data: "start",
      type: "start",
    },
    type: "trigger-event",
  },
  {
    detail: {
      blockedEvents: [
        "cold",
      ],
      logicStrands: [
        {
          priority: 1,
          strandName: "addHot",
        },
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
        {
          priority: 0,
          strandName: "start",
        },
      ],
      requestedEvents: [
        {
          data: "start",
          priority: 0,
          type: "start",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      data: "start",
      type: "start",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "cold",
      ],
      logicStrands: [
        {
          priority: 3,
          strandName: "mixHotCold",
        },
        {
          priority: 1,
          strandName: "addHot",
        },
        {
          priority: 2,
          strandName: "addCold",
        },
      ],
      requestedEvents: [
        {
          priority: 1,
          type: "hot",
        },
        {
          priority: 2,
          type: "cold",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "hot",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "hot",
      ],
      logicStrands: [
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
        {
          priority: 1,
          strandName: "addHot",
        },
      ],
      requestedEvents: [
        {
          priority: 2,
          type: "cold",
        },
        {
          priority: 1,
          type: "hot",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "cold",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "cold",
      ],
      logicStrands: [
        {
          priority: 1,
          strandName: "addHot",
        },
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
      ],
      requestedEvents: [
        {
          priority: 1,
          type: "hot",
        },
        {
          priority: 2,
          type: "cold",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "hot",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "hot",
      ],
      logicStrands: [
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 1,
          strandName: "addHot",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
      ],
      requestedEvents: [
        {
          priority: 2,
          type: "cold",
        },
        {
          priority: 1,
          type: "hot",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "cold",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "cold",
      ],
      logicStrands: [
        {
          priority: 1,
          strandName: "addHot",
        },
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
      ],
      requestedEvents: [
        {
          priority: 1,
          type: "hot",
        },
        {
          priority: 2,
          type: "cold",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "hot",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "hot",
      ],
      logicStrands: [
        {
          priority: 2,
          strandName: "addCold",
        },
        {
          priority: 3,
          strandName: "mixHotCold",
        },
      ],
      requestedEvents: [
        {
          priority: 2,
          type: "cold",
        },
      ],
    },
    type: "state-snapshot",
  },
  {
    detail: {
      type: "cold",
    },
    type: "selected-event",
  },
  {
    detail: {
      blockedEvents: [
        "cold",
      ],
      logicStrands: [
        {
          priority: 3,
          strandName: "mixHotCold",
        },
      ],
      requestedEvents: [
      ],
    },
    type: "state-snapshot",
  },
]
`;
